[core][gcs] Add storage namespace to redis storage in GCS.

iycheng opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
To enable one storage be able to be shared by multiple ray clusters, a special prefix is added to isolate the data between clusters: "<EXTERNAL_STORAGE_NAMESPACE>@"

The namespace is given by an os environment: `RAY_external_storage_namespace` when start the head: `RAY_external_storage_namespace=1234 ray start --head`

This flag is very important in HA GCS environment. For example, in ray serve operator, when the operator tries to bring up a new one, it's hard to just start a new db, but it's relatively easy to generate a new cluster id.
Another example is that, the user might only be able to maintain one HA Redis DB, and the namespace enable the user to start multiple ray clusters which share the same db.

This config should be moved to storage config in the future once we build that.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

iycheng commented:
@scv119 @mwtian do you have other comments?

iycheng commented:
@mwtian Thanks for the review.
I don't have strong opinion about this, but I think the direction we go should depend on the actual use case.

For 2), I'm not sure whether it's useful or just adds extra complexity. Any practical reason you think we should do this?

For 3) I am also not sure about this. Right now, we only have KubeRay trying to use this.

rkooo567 commented on src/ray/common/ray_config_def.h:
```diff
@@ -612,6 +612,10 @@ RAY_CONFIG(bool, scheduler_avoid_gpu_nodes, true)
 
+/// The ray cluster id. Different cluster should have unique ID.
+/// This fields is used to isolate data stored in DB.
+RAY_CONFIG(std::string, cluster_id, "default")
```
Isn't the default supposed to be an unique id if the different cluster should have unique id?

iycheng commented on src/ray/gcs/store_client/redis_store_client.cc:
```diff
@@ -17,23 +17,143 @@
+
+std::string GenKeyRedisMatchPattern(const std::string &cluster_id,
+                                    const std::string &table_name) {
+  return absl::StrCat(EscapeMatchPattern(cluster_id),
```
resolved offline.

iycheng commented:
@mwtian @rkooo567 @scv119 it's ready for another review. Really appreciate the discussion.

mwtian commented on src/ray/gcs/gcs_server/gcs_kv_manager.cc:
```diff
@@ -22,37 +22,45 @@
     return parts[1];
   }
-  return key;
+  return std::string(view.begin(), view.end());
```
^^?

mwtian reviewed (approved):
LGTM on the API, based on the off-thread discussion. I did not check the moved code in `src/ray/gcs/store_client/redis_store_client.cc`. Not sure if the test failures are related.

And just to confirm, have we removed all other places outside of KV where Ray writes to Redis?

iycheng commented:
> LGTM on the API, based on the off-thread discussion. I did not check the moved code in `src/ray/gcs/store_client/redis_store_client.cc`. Not sure if the test failures are related.
> 
> 
> 
> And just to confirm, have we removed all other places outside of KV where Ray writes to Redis?

Let me check the test failure.

Yes, we moved given that we have run ray without Redis for a while already, where all data are written to memory. This only change the memory access. I also have another PR to use Redis backend without pip install Redis.
